### PR TITLE
[COLAB-2495] Bugfix/colab 2495 profile jsp exception

### PR DESCRIPTION
**Pull request by cwoebker**
Wasn't sure what the best location would be to put the script. Let me know, and I will move it.

---
This change is [Reviewable](https://reviewable.io/reviews/cci-mit/xcolab/105)

### CODE REVIEW

File: view/src/main/webapp/WEB-INF/tags/formatDateTime.tagx [hunk 1]
```diff
@@ -10,6 +10,8 @@
     <jsp:useBean id="date" class="java.util.Date"/>
     <c:set target="${date}" property="time" value="${millis}"/>
 
+    <!-- TODO: need to check if ${date} is zero. COLAB-2495 Creation date could be Null. -->
```
Comment (jobachhu) [inline view/src/main/webapp/WEB-INF/tags/formatDateTime.tagx]: For new TODOs, we always add an issue code. If there's no issue about this yet, you can add it on Jira and then change it to `TODO COLAB-XXXX:`.

Is this still required though? The creation date should not be allowed to be null, right?

File: sql/deployments/deployment-2018-01-16-COLAB-2495.sql [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# Display all entries that have the issue
+SELECT * FROM xcolab.members_Member WHERE members_Member.status IS  NULL ;
+
+# Display entries without a createDate and the replacement createDate (DryRun)
+SELECT T1.id_, T1.screenName, T1.createDate, T2.createDate
+FROM xcolab.members_Member T1, xcolab.members_Member T2
+WHERE T1.id_ = T2.id_ + 1 AND T1.createDate IS NULL;
```
Comment (jobachhu) [inline sql/deployments/deployment-2018-01-16-COLAB-2495.sql]: It's good to have the SQL statements for reference, but I would comment them out so they're not run when we pipe this file to mysql.

File: sql/deployments/deployment-2018-01-16-COLAB-2495.sql [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+# Update status fields
+UPDATE xcolab.members_Member SET status=0 WHERE status IS NULL;
+UPDATE xcolab.members_Member SET autoRegisteredMemberStatus=0 WHERE autoRegisteredMemberStatus IS NULL;
```
Comment (jobachhu) [inline sql/deployments/deployment-2018-01-16-COLAB-2495.sql]: Please remove the database name (`xcolab.`) from the queries, as they may not be the same for everyone.

File: sql/deployments/deployment-2018-01-16-COLAB-2495.sql [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+# Update createDate field
+UPDATE xcolab.members_Member AS T1
+JOIN xcolab.members_Member AS T2
+    ON T1.id_-1 = T2.id_
+SET T1.createDate = T2.createDate
+WHERE T1.createDate IS NULL;
```
Comment (jobachhu) [inline sql/deployments/deployment-2018-01-16-COLAB-2495.sql]: Did this query populate all createDate fields when you ran it locally?

Two potential issues:
1. The ids range might not be contiguous everywhere.
2. If two consecutive createDates are null, the order matters. Is this dependency resolved correctly?

One potential solution could be to find the closest member with a create date in a subquery like this:
```sql
SELECT id_, ABS( id_ - myId ) AS distance
FROM members_Member
WHERE createDate is not null
ORDER BY distance
LIMIT 1
```

File: view/src/main/webapp/WEB-INF/tags/formatDateTime.tagx (hunk 1 above)
Comment (cwoebker) [reply view/src/main/webapp/WEB-INF/tags/formatDateTime.tagx]: Exactly, this was a note to myself earlier on. It's a design decision I guess. The advantage of not doing the check here is that we are more likely to find any issues with the database. :-)

File: sql/deployments/deployment-2018-01-16-COLAB-2495.sql (hunk 4 above)
Comment (cwoebker) [reply sql/deployments/deployment-2018-01-16-COLAB-2495.sql]: The script did work for all of the rows that I encountered. So 1 and 2 were not an issue for this one. I can however improve the script to account for those issues if you want.

Comment (jobachhu) [reply sql/deployments/deployment-2018-01-16-COLAB-2495.sql]: If it worked for all instances, then I think it's good enough - no need to over-engineer it :)

File: view/src/main/webapp/WEB-INF/tags/formatDateTime.tagx (hunk 1 above)
Comment (jobachhu) [reply view/src/main/webapp/WEB-INF/tags/formatDateTime.tagx]: I agree, let's keep it this way and not add the check - this way we'll fail early if something is wrong in the DB.

**Comment by cwoebker** [comment]
Should be done now. :-)

**Comment by cwoebker** [comment]
Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/deployments/deployment-2018-01-16-COLAB-2495-fix-member-null-fields.sql, line 7 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/105#-L3-b07UO4EkZwgVXF4p-r1-7:-L4729JJToklcnVOlOs9:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/28efa7fbad525db0722991d3a5c76f07f03e2ccc/sql/deployments/deployment-2018-01-16-COLAB-2495-fix-member-null-fields.sql#L7)):*
<details><summary><i>Previously, jobachhu (Johannes Bachhuber) wrote…</i></summary><blockquote>

It's good to have the SQL statements for reference, but I would comment them out so they're not run when we pipe this file to mysql.
</blockquote></details>

Done.

---

*[sql/deployments/deployment-2018-01-16-COLAB-2495-fix-member-null-fields.sql, line 11 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/105#-L3-b07UO4EkZwgVXF4p-r1-11:-L472GnECUtDHVUw7ZOC:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/28efa7fbad525db0722991d3a5c76f07f03e2ccc/sql/deployments/deployment-2018-01-16-COLAB-2495-fix-member-null-fields.sql#L11)):*
<details><summary><i>Previously, jobachhu (Johannes Bachhuber) wrote…</i></summary><blockquote>

Please remove the database name (`xcolab.`) from the queries, as they may not be the same for everyone.
</blockquote></details>

Done.

---

*[sql/deployments/deployment-2018-01-16-COLAB-2495-fix-member-null-fields.sql, line 10 at r3](https://reviewable.io:443/reviews/cci-mit/xcolab/105#-L3oSJNf7E5RqOZUxiH6-r3-10:-L472FmQV6iH32bye2vC:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/2b6c181a628ec7908a60dcecc0ef437b41899ff0/sql/deployments/deployment-2018-01-16-COLAB-2495-fix-member-null-fields.sql#L10)):*
<details><summary><i>Previously, jobachhu (Johannes Bachhuber) wrote…</i></summary><blockquote>

This query still has the database name prefixed to the table name. Please remove the `xcolab.` prefix so it's always run in the currently selected database.

In its current form, the script wouldn't work if the database has a different name, e.g. on our homolog server (we shouldn't assume every one uses the same name).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cci-mit/xcolab/105)*

**Comment by jobachhu** [comment]
Reviewed 1 of 3 files at r2, 1 of 2 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cci-mit/xcolab/105)*

**Comment by jobachhu** [comment]
Looks great, thanks! 👍

**Pull request merged**